src/particles/distribution/Waterbag.H

```diff
@@ -0,0 +1,75 @@
+/* Copyright 2021 Ji Qiang, Axel Huebl
+ *
+ * This file is part of ImpactX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef IMPACTX_DISTRIBUTION_WATERBAG
+#define IMPACTX_DISTRIBUTION_WATERBAG
+
+#include <AMReX_Random.H>
+#include <AMReX_REAL.H>
+
+#include <cmath>
+
+
+namespace impactx
+{
+namespace distribution
+{
+    struct Waterbag
+    {
+        /** A Waterbag distribution
+         *
+         * return 1 particle coordinates sampling from a 6D Waterbag distribution
+         * with mean = 0, std = 1.
+         */
+        Waterbag() = default;
+
+        /** Return 1 6D particle coordinate
+         *
+         * @param x particle position in x
+         * @param y particle position in y
+         * @param t particle position in t
+         * @param px particle momentum in x
+         * @param py particle momentum in y
+         * @param pt particle momentum in t
+         */
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        void operator() (
+                amrex::ParticleReal & x,
+                amrex::ParticleReal & y,
+                amrex::ParticleReal & t,
+                amrex::ParticleReal & px,
+                amrex::ParticleReal & py,
+                amrex::ParticleReal & pt,
+                amrex::RandomEngine const& engine) const {
+
+            using namespace amrex::literals;
+
+            amrex::ParticleReal r1=0.0, r2=0.0, r3=0.0;
+            amrex::ParticleReal x1=0.0, x2=0.0, x3=0.0;
+
+            for(;;)
+            {
+                r1 = 2_prt * amrex::Random(engine) - 1.0_prt;
+                r2 = 2_prt * amrex::Random(engine) - 1.0_prt;
+                r3 = 2_prt * amrex::Random(engine) - 1.0_prt;
+                x1 = 2_prt * amrex::Random(engine) - 1.0_prt;
+                x2 = 2_prt * amrex::Random(engine) - 1.0_prt;
+                x3 = 2_prt * amrex::Random(engine) - 1.0_prt;
+                if(r1*r1 + r2*r2 + r3*r3 + x1*x1 + x2*x2 + x3*x3 < 1.0_prt) break;
+            }
```

Comment:
Just as a general note:
A rejection method will not be very performant on GPU, since we will see a lot of thread divergence here.

If we can express this as a transformation of a uniform or standard distribution (for which be both have functions as well to pull random numbers from), then this would be better.

Reply:
I have to think about this a bit.

Reply:
@cemitch99 added on Slack:
> This can be done.  I have some Fortran code for KV-like and waterbag distributions that use a transformation based on 1D uniform sampling--I can share, discuss with Ji.